ONNX Interpolate Add Scales Params

lara-hdr opened the pull request:
Fix for : https://github.com/pytorch/pytorch/issues/27176

Commit:
Add Scales Attribute to Upsample_Nearest Ops

Commit:
add linear and cubic

Commit:
add update to test

Commit:
rebase

gchanan commented:
@houseroad can you comment on the backwards compatibility concerns here?

lara-hdr committed:
Update UpSampleNearest3d.cu

lara-hdr committed:
Update UpSampleNearest2d.cu

lara-hdr committed:
Update UpSampleNearest1d.cu

houseroad reviewed (changes requested):
Adding new parameter without default value is backward incompatible change. This will break existing models. So could you add a default value for it?

Commit:
add default values

Commit:
pull

Commit:
missing files

Commit:
fix ci failure

Commit:
rebase

lara-hdr commented:
Hi,
While making this change and testing against ONNX I discovered the following.

Let's take the following example:
The user calls torch.nn.functional(input, mode=‘bilinear’, scale_factor=2.3) with an input of size (1, 2, 4, 4).

![interpolate_explanation](https://user-images.githubusercontent.com/11823865/67522041-d7d71080-f660-11e9-9467-a8e70731a1eb.png)

++ Today in PyTorch Master:

- The scales are converted to output_size in torch.nn.functional here https://github.com/pytorch/pytorch/blob/e9a91756cddbb41d584363528b18169e529007a3/torch/nn/functional.py#L2474.
The output_size takes the value of (1, 2, 9, 9).
- The aten op aten::upsample_bilinear2d is called with the computed output_size (1, 2, 9, 9).
- The scales are computed again in the op (here: https://github.com/pytorch/pytorch/blob/420b37f3c67950ed93cd8aa7a12e673fcfc5567b/aten/src/ATen/native/UpSample.h#L133 (this actually computes 1/scale)), using the output_size this time, which results in a different scale as the one given by the user; in this case (1, 1, 2.25, 2.25) rather than (1, 1, 2.3, 2.3).

We follow a similar logic in the ONNX operator.

++ In this PR:

- We are introducing new parameters as inputs of the aten op aten::upsample_bilinear2d with the values of the scales given by the user (in this case 2.3).
- However, we are not changing the aten op's logic, which is still computing the scales based on the computed output_size rather than using the one given by the user.
(so the used scales are (1, 1, 2.25, 2.25) rather than (1, 1, 2.3, 2.3)).

- On ONNX's side, we are now directly using the scales given as inputs. 
(So the scales are (1, 1, 2.3, 2.3), as given by the user).

As explained, ONNX and PyTorch are now not computing the same output since they are relying on different scales.

However, now that the scales are available as inputs of the aten ops, using them directly in the aten ops could benefit them as well, by avoiding mismatches between the computed scales and the ones given in the api call.
We are suggesting to modify the logic in https://github.com/pytorch/pytorch/blob/420b37f3c67950ed93cd8aa7a12e673fcfc5567b/aten/src/ATen/native/UpSampleBilinear2d.cpp#L44 (and the other aten upsample ops), to use the scales given as inputs (when they are available), instead of computing them based on the previously computed output size.

Any thoughts on this proposal?
@houseroad @gchanan @spandantiwari

spandantiwari commented:
@lara-hdr - thanks for the detailed description. 

> - We are introducing new parameters as inputs of the aten op aten::upsample_bilinear2d with the values of the scales given by the user (in this case 2.3).
> - However, we are not changing the aten op's logic, which is still computing the scales based on the computed output_size rather than using the one given by the user.
> (so the used scales are (1, 1, 2.25, 2.25) rather than (1, 1, 2.3, 2.3)).
> - On ONNX's side, we are now directly using the scales given as inputs.

The problem fundamentally seems to be that even if `scales` is provided as input by the user is not directly used in the computation in the lower ATEN op computation. I this PR you have added the argument to the ATEN api but the computation underneath does not use that argument. My recommendation would be to consider a minor update to the ATEN op to use the scales directly when they are provided as the input argument, in the lines that you pointed out. I feel that this would be the cleanest fix to this issue. 

@houseroad @gchanan - let us know what you think. Would it be OK to make the update to the ATEN op to work directly with `scales`, when provided?

Commit:
rebase and resolve conflicts

lara-hdr commented:
@houseroad, @gchanan did you get a chance to look into the comments above?

houseroad reviewed (commented):
I agree with @spandantiwari. Can we decide whether we use scale or not inside ATen function?

spandantiwari commented:
@gchanan - Could you share your thoughts on this?

Commit:
use scales directly in aten ops (cpu)

Commit:
update aten gpu

Commit:
revert changes on nearest gpu - CI failure

lara-hdr commented:
@pytorchbot rebase this please

Commit:
Merge branch 'master' of https://github.com/lara-hdr/pytorch into lahaidar/improve_interpolate_onnx

lara-hdr committed:
Merge branch 'master' of https://github.com/lara-hdr/pytorch into lahaidar/improve_interpolate_onnx

houseroad commented:
pytorchbot stopped working, please manually do the rebase

Commit:
-

Commit:
rename var

BowenBao commented on aten/src/ATen/native/cuda/UpSampleNearest1d.cu:
```diff
@@ -22,7 +22,7 @@ __global__ void upsample_nearest1d_out_frame(
     size_t dst_dim_w,
     scalar_t* output) {
-  int dst_idx = blockIdx.x * blockDim.x + threadIdx.x;
+  size_t dst_idx = blockIdx.x * blockDim.x + threadIdx.x;
```
why switch to `size_t`?

lara-hdr replied:
I just reverted UpSampleNearest1/2/3d.cu because of some CI failures. 
It seems like this was changed in another PR, so by reverting to an older commit in this PR I lost the merge.

lara-hdr replied:
(same for the other "ditto")

BowenBao commented on aten/src/ATen/native/cuda/UpSampleNearest2d.cu:
```diff
@@ -163,17 +164,8 @@ static void upsample_nearest2d_out_cuda_template(
   int grid_y = cuda::ATenCeilDiv(output_height, block_y);
-  int grid_z = std::min<int>(
-      maxGridSize[2], cuda::ATenCeilDiv(nc, block_z * 4));
+  int grid_z = cuda::ATenCeilDiv(nc, block_z * 4);
```
What is this change doing? Is this related to scales being added as parameter?

BowenBao commented on aten/src/ATen/native/cuda/UpSampleNearest2d.cu:
```diff
@@ -244,9 +238,6 @@ static void upsample_nearest2d_backward_out_cuda_template(
       at::cuda::getCurrentDeviceProperties()->maxThreadsPerBlock, MAX_THREADS)};
   dim3 gdim{cuda::ATenCeilDiv(n, bdim.x)};
-  // safe check for int32 indexing; implicitly restrict launch config for kernel
-  TORCH_CHECK(grad_input.numel() <= std::numeric_limits<int32_t>::max());
```
ditto

BowenBao commented on aten/src/ATen/native/cuda/UpSampleNearest3d.cu:
```diff
@@ -163,9 +166,6 @@ static void upsample_nearest3d_out_cuda_template(
       at::cuda::getCurrentDeviceProperties()->maxThreadsPerBlock, MAX_THREADS)};
   dim3 gdim{cuda::ATenCeilDiv(n, bdim.x)};
-  // safe check for int32 indexing; implicitly restrict launch config for kernel
-  TORCH_CHECK(output.numel() <= std::numeric_limits<int32_t>::max());
```
ditto

BowenBao commented on aten/src/ATen/native/cuda/UpSampleNearest3d.cu:
```diff
@@ -242,9 +245,6 @@ static void upsample_nearest3d_backward_out_cuda_template(
       at::cuda::getCurrentDeviceProperties()->maxThreadsPerBlock, MAX_THREADS)};
   dim3 gdim{cuda::ATenCeilDiv(n, bdim.x)};
-  // safe check for int32 indexing; implicitly restrict launch config for kernel
-  TORCH_CHECK(grad_input.numel() <= std::numeric_limits<int32_t>::max());
```
ditto

BowenBao commented on test/onnx/test_pytorch_onnx_onnxruntime.py:
```diff
@@ -502,68 +502,80 @@ def forward(self, x):
+        model = MyModel(mode, use_size, is_upsample, align_corners)
         self.run_test(model, x)
 
     # TODO: Enable bicubic, linear1d and linear3d when implemented in ORT
```
Are these tests enabled now?

lara-hdr replied:
yes! cubic is the only one that is still disabled

Commit:
fix nearest gpu

Commit:
merge conflicted files

houseroad commented:
Rebase again?

lara-hdr committed:
Merge branch 'lahaidar/improve_interpolate_onnx' of https://github.com/lara-hdr/pytorch into lahaidar/improve_interpolate_onnx

lara-hdr committed:
merge

lara-hdr committed:
Merge branch 'master' of https://github.com/lara-hdr/pytorch into lahaidar/improve_interpolate_onnx

lara-hdr commented:
@houseroad just did but I am still getting a failure in test_interpolate_nearest_scale_1/2/3d_cuda. I'm looking into it.

lara-hdr committed:
lnit

lara-hdr committed:
move compute_scales_value

lara-hdr committed:
fix scale in backward

lara-hdr committed:
-

lara-hdr committed:
merge conflict

lara-hdr committed:
quantized op

Commit:
rebase and merge conflict

Commit:
Merge branch 'lahaidar/improve_interpolate_onnx' of https://github.com/lara-hdr/pytorch into lahaidar/improve_interpolate_onnx

BowenBao commented on torch/onnx/symbolic_helper.py:
```diff
@@ -320,6 +343,21 @@ def _interpolate_get_scales_and_mode(g, input, size, scale_factor, mode , align_
+    else:
+        align_corners = args[0]
+        scales = args[1:]
+    if len(scales) == 1:
```
Maybe change to `return scales, align_corners`

BowenBao commented on torch/onnx/symbolic_opset9.py:
```diff
@@ -733,23 +733,46 @@ def replication_pad(g, input, padding):
+def _interpolate1d(name, dim, interpolate_mode):
+    def symbolic_fn(g, input, output_size, *args):
+        scales_1, align_corners = sym_help.get_interpolate_attributes(interpolate_mode, args)
+        scales = sym_help._interpolate_get_scales_if_available(g, [scales_1])
```
With the above comment update on `get_interpolate_attributes`. It seems the duplication of explicitly writing methods for 1d, 2d, 3d could be removed by 
```
        scales, align_corners = sym_help.get_interpolate_attributes(interpolate_mode, args)
        scales = sym_help._interpolate_get_scales_if_available(g, scales)
```

Commit:
Merge branch 'master' of https://github.com/lara-hdr/pytorch into lahaidar/improve_interpolate_onnx

Commit:
refactor fctions in exporter

Commit:
lint

houseroad commented:
Nov 16 02:44:22 test/onnx/test_pytorch_onnx_onnxruntime.py::TestONNXRuntime_opset11_IRv4::test_interpolate_upsample FAILED [ 81%]

lara-hdr commented:
that's weird, I didn't see this failure locally.. and the test passed in previous commits.. 
It could be a precision error, I'll try relaxing the precision

lara-hdr committed:
diable bicubic until new ORT nightly

spandantiwari commented on aten/src/ATen/native/UpSample.h:
```diff
@@ -111,11 +111,22 @@ static inline void upsample_3d_shape_check(
+    int64_t input_size,
+    int64_t output_size) {
+      return (scale > 0.)
+          ? static_cast<scalar_t>(1. / static_cast<scalar_t>(scale))
```
It may be better to do this division is double-precision since scale is coming in as a double and cast it to float just once. So something like:
`? static_cast<scalar_t>(1.0 / scale)`

spandantiwari commented on aten/src/ATen/native/UpSample.h:
```diff
@@ -111,11 +111,22 @@ static inline void upsample_3d_shape_check(
+    int64_t output_size) {
+      return (scale > 0.)
+          ? static_cast<scalar_t>(1. / static_cast<scalar_t>(scale))
+          : (static_cast<scalar_t>(input_size) / output_size);
```
`input_size` and `output_size` are integers. Should we cast them as float first before division to get floating point division and output. That will be more accurate for scale.

lara-hdr replied:
Actually the division is between static_cast<scalar_t>(input_size) and output_size. 
So input_size is converted to scalar_t (which would be float or double), then divided by output_size (int).
(The division is between a float/double and an int, so the compiler would convert output_size to scalar_t as well.)

spandantiwari commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -6035,171 +6035,171 @@
     CUDA: upsample_linear1d_out_cuda
 
-- func: upsample_linear1d(Tensor self, int[1] output_size, bool align_corners) -> Tensor
+- func: upsample_linear1d(Tensor self, int[1] output_size, bool align_corners, float scales_1=-1.0) -> Tensor
```
nit: `float scales_1=-1.0f` for more readability.

lara-hdr replied:
this is an invalid notation in yaml

spandantiwari commented on torch/csrc/jit/register_prim_ops.cpp:
```diff
@@ -3400,16 +3400,36 @@ at::Tensor interpolate(
     }
   }
 
+  float scale_factors_1 = -1;
```
nit: `-1.0f`

spandantiwari commented on torch/csrc/jit/register_prim_ops.cpp:
```diff
@@ -3400,16 +3400,36 @@ at::Tensor interpolate(
+  float scale_factors_2 = -1;
+  float scale_factors_3 = -1;
+  if (scale_factors.isDouble()) {
+    scale_factors_1 = (float) scale_factors.toDouble();
```
Not clear on why we need to convert explicitly to double using `.toDouble` if it is already a double (as checked in the line above). 
Also, for casting to float, can we use the modern `static_cast<float>`?

lara-hdr replied:
scale_factors is not a double, it is an instance of IValue.
toDouble() is a function of the class that returns the value of the double.

spandantiwari commented on torch/onnx/symbolic_helper.py:
```diff
@@ -269,6 +269,35 @@ def _interpolate_size_to_scales(g, input, output_size, dim):
+    if not available_scales:
+        return None
+
+    from torch.onnx.symbolic_opset9 import unsqueeze
```
Seems weird that `symbolic_helper` is importing code from a specific symbolic opset.

lara-hdr replied:
I agree that this is not ideal.
Do you have an idea of how to avoid doing that?

One way would be to have a helper function for unsqueeze in symbolic_helper (something like unsqueeze_helper()), that will call the right unsqueeze depending on the opset version (today the only one is the one in symbolic_opset9). 
Would that be better?

spandantiwari replied:
I think it would be. Any common code can be pulled into helper, and symbolic can be downstream from it.

spandantiwari commented on torch/onnx/symbolic_opset11.py:
```diff
@@ -48,25 +48,37 @@ def pixel_shuffle(g, self, upscale_factor):
+                        empty_tensor,  # roi only takes effect whith coordinate_transformation_mode="tf_crop_and_resize"
+                        scales,  # scales is not needed since we are sending out_size
+                        coordinate_transformation_mode_s=coordinate_transformation_mode,
+                        cubic_coeff_a_f=-0.75,  # only valid when mode="cubic"
```
nit: `...=-0.75f`.

lara-hdr replied:
this is python code

spandantiwari replied:
My bad.

lara-hdr committed:
code readibility float/double

Commit:
add helper for unsqueeze

Commit:
lint

Commit:
merge conflicts

spandantiwari reviewed (approved):
LGTM.

ailzhang reviewed (changes requested):
Thanks for the fix! 
Talked to @houseroad offline about the issue, we think this PR could be written without BC breaking change. 
First it's clear that the BC-breaking change of this PR comes from changes in `area_pixel_compute_scale` (and its variants), that the logic now depend on whether `scale > 0`. Following the example in comments above, here the computed scale will be `2.3` instead of `2.25`. But in the interpolation computation, when you scale an image of `(4, 4)` to `(9, 9)`, the real scale that should be used in the interpolation is `2.25`, but not `2.3`. Thus I think the change produces different results than before and it's wrong scale to use in implementation.
For the purpose of fixing ONNX use case, you can simply passing `scales` into C++ functions, but just leave them not used in the ATen implementation. That way ONNX should get the correct `scales` and this PR will no longer be BC breaking. 
Also when you make the change, it'll be great to add a note in comment to explain that why we added `scale` and why it is not used. A link to this PR is also good to have. 
Please correct me if I understand the problem wrong.

lara-hdr commented:
@ailzhang @houseroad, this was the state of the PR as of Oct 25th (passing the scales to the ATen op but only use them in ONNX).
This is when we discovered the mismatch between ONNX and PyTorch; the above comment (https://github.com/pytorch/pytorch/pull/28324#issuecomment-546088502) was to discuss the mismatch in behavior/results, and discuss how the ATen ops can now use the scales directly (to avoid loosing precision by converting them to output size and back to scales), and decide whether or not to implement the behavior we have today in this PR.

ailzhang commented:
@lara-hdr Yea I see the proposal above. Hmmm but I doubt if we should change pytorch behavior here. When you resize `(4, 4)` to `(9, 9)`, although user specified factor is `2.3`, the real factor is `2.25`. So the number to use in interpolation calculation is `2.25` instead of `2.3`. 

After talking to @houseroad , IIRC `2.3` is useful to export the correct user defined behavior. With that said, we should move the `output size` calculation into c++, so that we pass both `scale, output_size` to C++. In that case, JIT & ONNX knows whether user wanted fixed `scale` or fixed `output_size`. 

But on the other hand, in ATen implementation, once we calculated `output_sizes` (either based on user specified `scale` or user specified `output_sizes`), the user specified scale should no longer be used in any calculation. IMHO `area_pixel_compute_scale` behavior should remain unchanged. 
Please correct me if I'm wrong. Thanks!

lara-hdr commented:
@ailzhang, @houseroad, I don't agree with some of the affirmations.

With the above example (input_size = (4,4), scales = 2.3), the interpolation should use the scale 2.3 when computing the output.

Currently in master, the output_size (9, 9) is computed (output_size = floor(input_size * scales)), and the ATen op is called with output_size=(9, 9), the scales are then computed from the output size, which results in 2.25 instead of 2.3.
But 2.3 is what the user specified and is what should be used.
2.25 is an approximation and using it will result in a wrong output.

Saying that 2.25 should be used when the user is specifying 2.3, is saying that an interpolation with scales = 2.25 and an interpolation with scales = 2.3 are always equivalent for an input of of size (4, 4), which is inaccurate.

In other words, the current calculations in master are using 
`scales = floor(input_size * scales) / input_size, `
which is mathematically incorrect (now that we have the scales available in the ATen op, we should simply use scales = scales).

With the changes we are suggesting, we will be directly using the scales specified by the user. 
The output of the operator will indeed differ in some cases from the old behavior, but this correction will allow the results to be more accurate.

ailzhang commented:
Hey @lara-hdr, I talked to @ppwwyyxx about this issue that we think this is an edge case(user could make it more clear by calculating the `output_sizes` and pass in those instead of an obscure `scale`) that currently it's undefined behavior. 

As reference TF only allows passing in `output_sizes`, while opencv (which our interpolate methods have been matching) has a similar option. And here's the comparison on master:
```
In [9]: import cv2
   ...: import torch
   ...: import torch.nn.functional as F
   ...:
   ...: t = torch.rand(1, 17, 4, 4)
   ...:
   ...: pt_result = F.interpolate(t, scale_factor=2.3, mode="bicubic", align_corners=False)
   ...: cv_result = torch.from_numpy(cv2.resize(t.squeeze(0).permute(1, 2, 0).numpy(), None, fx=2.3, fy=2.3, interpolation=cv2.INTER_CUBIC)).permute(2, 0, 1).unsqueeze(0)
   ...:
   ...: print(pt_result.shape)
   ...: print("max abs diff:", (pt_result - cv_result).abs().max())
   ...: print("MSE:", ((pt_result - cv_result) ** 2).mean())
   ...:
   ...:
   ...:
torch.Size([1, 17, 9, 9])
max abs diff: tensor(0.0897)
MSE: tensor(0.0007)
```
If you could run this code snippet against this PR, and let us know whether it improves the max abs difference. We can then safely conclude whether to fix it in a BC breaking way or not. Thanks!

lara-hdr commented:
@ailzhang I ran the code and with this PR. The output is:

```
torch.Size
([1, 17, 9, 9])
max abs diff: tensor(1.6689e-06)
MSE: tensor(1.2435e-13)
```

spandantiwari commented:
> As reference TF only allows passing in `output_sizes`, while opencv (which our interpolate methods have been matching) has a similar option. And here's the comparison on master:
> 
It is quite common for interpolate and resize APIs in different frameworks to have an input argument for scale factor(s). Opencv has that option too ( [cv.resize2](https://docs.opencv.org/2.4/modules/imgproc/doc/geometric_transformations.html?highlight=resize#cv2.resize) ), where you can specify scale factor and not output sizes . In fact, going by their documentation below, I believe the scale factor argument gets preference.
` "Note that the initial dst type or size are not taken into account. Instead, the size and type are derived from the src,``dsize``,``fx`` , and fy ." ` 

There are other examples such as [imresize](https://www.mathworks.com/help/images/ref/imresize.html) in MATLAB. 

If a scale factor is provided as input to PyTorch API, users would expect it be honored as closely as possible in the implementation, and the current approach of converting it to output size in the Python layer and then converting back to scale factor in C++ ATen code, is loosing some precision and is not the most accurate implementation. In fact, the reason this change is being proposed is because, after our investigation, we found that it is not matching ONNX backend implementations which use the scale factor directly. 

Therefore, at this point, this seems like an improved implementation and should be treated as such.

ailzhang commented on torch/nn/functional.py:
```diff
@@ -2507,37 +2507,44 @@ def _output_size(dim):
             align_corners = False
 
     if input.dim() == 3 and mode == 'nearest':
```
warning example:https://github.com/pytorch/pytorch/blob/effcc398c447ce2aa6752354d64ef3d7ac5f004b/torch/nn/functional.py#L1599
note example: https://github.com/pytorch/pytorch/blob/effcc398c447ce2aa6752354d64ef3d7ac5f004b/torch/nn/functional.py#L2678

ailzhang reviewed (approved):
Given that this PR brings us closer to opencv result, I think it's good to add it. Although might be minor, this is a BC-breaking change that we need to handle carefully. Thus I have a few minor comments to make it easier for users/devs to understand. 
Also since this changes eager frontend behavior as well, please add a "expect" test similar to https://github.com/pytorch/pytorch/blob/effcc398c447ce2aa6752354d64ef3d7ac5f004b/test/test_nn.py#L7775. 
Thanks!

lara-hdr committed:
add expect test; add notes; add warning

lara-hdr commented:
@ailzhang thanks a lot, I addressed the comments in the last commit.
Also, as I mentioned to houseroad, I see a failure in pytorch_xla_linux_xenial_py3_6_clang7_build that I don't see locally.
"AtenXlaType function missed override".
Do you know why this error is happening?

lara-hdr committed:
fix nearest3d

houseroad commented:
@ailzhang what's the plan for xla build :-)

ailzhang commented:
I’ll prepare a pr tmr afternoon. ;) we can plan to land it tmr/wed

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________
From: Lu Fang <notifications@github.com>
Sent: Monday, November 25, 2019 5:44:07 PM
To: pytorch/pytorch <pytorch@noreply.github.com>
Cc: Ailing Zhang <ailzhang@fb.com>; Mention <mention@noreply.github.com>
Subject: Re: [pytorch/pytorch] ONNX Interpolate Add Scales Params (#28324)


@ailzhang<https://github.com/ailzhang> what's the plan for xla build :-)

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/pytorch/pytorch/pull/28324?email_source=notifications&email_token=ABIBI6TJJBMIZERI37BNTF3QVR5OPA5CNFSM4JCNE6F2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEFENECY#issuecomment-558420491>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ABIBI6WUTLW2OLTOVU4YIUDQVR5OPANCNFSM4JCNE6FQ>.

ailzhang commented:
@lara-hdr I've add a pr on XLA side to fix the breakage. 
The clang-tidy failure might be real as well.

lara-hdr commented:
@ailzhang, the clang-tidy is a warning because of the use of magic-numbers instead of named variables.
But the use of those numbers was not introduced by this PR..
Should I still fix it?

ailzhang commented:
As I worked on the XLA part for this PR, XLA/TPU version of `resize` doesn't support `scale_factor` as I mentioned above, thus this PR would introduce a gap before Pytorch and Pytorch/XLA.

I'm not certain how beneficial it would be if we introduce this BC breaking change. I think @fmassa might give helpful input here. 

@lara-hdr To unblock you first, I think we can easily land the non-BC breaking part (that doesn't change scale value) if you split it to two PRs. What do you think?

lara-hdr commented:
@ailzhang, thanks for helping with the XLA part.
I see.. I could split the PR, but I am curious what would it take to support scale_factor in resize on XLA/TPU?

ailzhang commented:
@lara-hdr IIUC it'll require a change on XLA side to consider the scale_factor as well. 
Yea it will be much easier if we land the easy part, and focus on the BC discussion and changes that are blocking it on a separate PR. ;)  Thanks!

Commit:
merge

Commit:
Merge branch 'lahaidar/improve_interpolate_onnx' of https://github.com/lara-hdr/pytorch into lahaidar/improve_interpolate_onnx

lara-hdr commented:
@ailzhang I talked with @spandantiwari; it would be preferable to keep the PR with all the changes.
Merging only the non BC breaking changes of this PR won't unblock us since we won't be matching behaviors with PyTorch.
Could you share what it would take to support scale_factor in XLA/TPU and how much time would that require?

Commit:
add use_scale_factors param

Commit:
merge conflicts

lara-hdr commented:
@ailzhang @houseroad @fmassa, I am fixing the test error, but it would be great to get a first round of review since we have a short deadline, and I will be updating the PR with a fix asap.
Thanks!

Commit:
fix failing test

lara-hdr committed:
fix line-too-long failure

lara-hdr commented:
I fixed the test failure and tests should be passing now.
@ailzhang I am still getting the "AtenXlaType function missed override" error.

ailzhang commented:
FYI this is no longer a BC breaking change, but we should add it in the 1.4 release note to warning about default behavior change in 1.5. 
@lara-hdr Once this landed to master, please also open another PR against 1.4 branch.

lara-hdr commented:
Thanks a lot @ailzhang !!

houseroad reviewed (approved):
Finally, ready to ship

houseroad commented:
Final solution: We don't break the BC in this PR, but we introduce additional parameter, and warn users that we will introduce BC breaking change in next release (1.5). cc: @gchanan

houseroad commented:
FYI: this PR will break XLA CI on purpose.

eellison commented on torch/csrc/jit/register_prim_ops.cpp:
```diff
@@ -3382,16 +3389,75 @@ at::Tensor interpolate(
     }
   }
 
+  float scale_factors_1 = -1.0f;
```
Are these changes tested anywhere ?

lara-hdr replied:
there is a test_upsampling_use_scale_factor test in test_nn.py

eellison replied:
Yep but it doesn't look like that invokes the scripted function at all

lara-hdr replied:
@eellison in which file would you suggest adding a test?

eellison replied:
Next to all the tests here: https://github.com/pytorch/pytorch/blob/b8b6cee123343228fbb83e28b27bd0001393c173/test/test_jit.py#L16927

eellison reviewed (changes requested):
Doesn't look like `use_scale_factor` changes are tested at all for torchscript. It's unfortunate that we need two implementations... I'm going to look into seeing if we can compile `interpolate` now that we have overloaded functions in upcoming weeks.

Commit:
add script test in test_jit

lara-hdr commented:
@eellison I added tests in test_jit.py

Commit:
pull

lara-hdr committed:
nit

houseroad commented:
@eellison thanks a lot for suggesting adding test to cover the jit part. We would like to land this as soon as possible, since we are targeting enable this feature in release 1.4, and start sending the message to users. Could you please check if the tests are good enough or not?

houseroad commented:
@lara-hdr is the windows build failure related?

Commit:
split string

Commit:
Merge branch 'lahaidar/improve_interpolate_onnx' of https://github.com/lara-hdr/pytorch into lahaidar/improve_interpolate_onnx

lara-hdr commented:
@houseroad sorry it's a new failure that I hadn't notice.
It was a "string too long" error; I just pushed a commit to split the string, which should fix it.

Commit:
move adaptive_avg

eellison commented:
Taking a look now

eellison commented:
@ailzhang could you please review the autograd changes ? The test coverage for it was added since you last reviewed

zou3519 commented on torch/nn/functional.py:
```diff
@@ -2457,6 +2457,11 @@ def interpolate(input, size=None, scale_factor=None, mode='nearest', align_corne
+            directly used in the output computation, or can be used to compute the output size which
+            will later be used to infer new scales values.
+            If set to ``True``, scale_factors are used in the interpolation.
+            If set to ``False``, the output_size is computed and new scales are infered for the interpolation.
```
What happens when `use_scale_factor` is None?

lara-hdr replied:
The default value is False for now and will be changed to default to True in 1.5.0

lara-hdr replied:
@zou3519, for more context:
This PR introduces a change in the computation of interpolate, that would had been BC breaking without the added parameter 'use_scale_factor'. 

We had a meeting to discuss this change with @houseroad @ailzhang @fmassa @spandantiwari;
The outcome was to add a parameter (that default to False) so the change is not a breaking change in 1.4.0. 

So 'use_scale_factor' defaults to False in 1.4.0 to keep the old behavior, and we warn the users that the default behavior will change in version 1.5.0.

zou3519 replied:
Thanks for the context. If the default is `false` we should record in here that the default is false.

I'm having trouble understanding from the explanation in the documentation what exactly is going to be different -- does the value of the output tensor change depending on `use_scale_factor`?

lara-hdr replied:
Yes the behavior is different and affects the output.

Interpolate uses scale values to compute the output. 
Interpolate can be called with scale_factor param, which are the scales to use in the interpolation computation.

- with use_scale_factor = False (old behavior that we wish the deprecate):
The output_size is computed as : output_size = input_size * scale_factor
Then scales are computed later in the aten op : scales = output_size / input_size
With this behavior there is a loss in precision in the computed scales.


- with use_scale_factor = True (new desired behavior):
scale_factor is passed to the aten op and scales = scale_factor

The scales used in the interpolation being different, the output of interpolate will differ in certain cases, which is why it is important to explain this param in the doc and warn the user about the behavior change in 1.5.0.

zou3519 replied:
Okay, I see. It was not clear to me how the new scales could be different from the input scales (from your explanation they can be different due to floating point precision). That might be a good thing to note in the docs?

But at any rate, I understand the value of the new parameter, thank you for clarifying.

lara-hdr replied:
Great, thanks. Lmk if there are other suggestions, otherwise I will push a new commit with the doc updated as suggested.

zou3519 replied:
I read some more context from https://github.com/pytorch/pytorch/issues/27176. To clarify, is the output different only when doing ONNX export, or is it different when running pytorch python as well?

lara-hdr replied:
This issue was opened before we discovered the precision loss in the scales.
Now the output is also different in Pytorch when use_scale_factor is True

zou3519 commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -6117,171 +6117,171 @@
     CUDA: upsample_nearest3d_backward_out_cuda
 
-- func: upsample_nearest3d_backward(Tensor grad_output, int[3] output_size, int[5] input_size) -> Tensor
+- func: upsample_nearest3d_backward(Tensor grad_output, int[3] output_size, int[5] input_size, float scales_1=-1.0, float scales_2=-1.0, float scales_3=-1.0) -> Tensor
```
Two questions:
- Could we pass a `float[3]` instead? It's not really scalable to add new args
- Should these be zero-indexed instead of one-indexed?

lara-hdr replied:
This was my first try but adding float arrays in ative_functions.yaml is not supported

eellison replied:
How difficult would it be to add `float[]` in native_functions.yaml ? once we go with this then we can't undo it without breaking BC.

zou3519 commented on torch/nn/functional.py:
```diff
@@ -2473,6 +2478,14 @@ def interpolate(input, size=None, scale_factor=None, mode='nearest', align_corne
         See :class:`~torch.nn.Upsample` for concrete examples on how this
         affects the outputs.
 
+        When scale_factor are specified, if use_scale_factor=False,
```
"scale_factor is specified"

zou3519 commented on torch/nn/functional.py:
```diff
@@ -2473,6 +2478,14 @@ def interpolate(input, size=None, scale_factor=None, mode='nearest', align_corne
 
+        When scale_factor are specified, if use_scale_factor=False,
+        scale_factor is used to compute the output_size which will then
+        be used to infer new scales for the interpolation. This is the current
```
This reads like implementation detail to me: do we really need to tell users about this? I think what users should know about is how including this parameter changes the ONNX graph.

lara-hdr replied:
@zou3519, it is important to explain the difference in behavior since this parameter also affects PyTorch's interpolate output, to use the correct values of the scales when use_scale_factor=True, as explained in this message.

zou3519 replied:
I'm having trouble, from reading this description, understanding how this parameter affects the interpolate output. But maybe that is because I don't understand the interpolate operator.

zou3519 replied:
resolving based on https://github.com/pytorch/pytorch/pull/28324#discussion_r355661922

eellison commented on torch/nn/functional.py:
```diff
@@ -2486,6 +2499,17 @@ def _check_size_scale_factor(dim):
+            if use_scale_factor is None:
+                is_float_scale_factor = any(not float(scale).is_integer() for scale in _ntuple(dim)(scale_factor))
+                if is_float_scale_factor:
+                    warnings.warn("The default behavior for interpolate/upsample with float scale_factor will change "
```
Is there a particular reason we're not exposing `use_scale_factor` for upsample? (maybe bc they are deprecated, just curious).

lara-hdr replied:
yes. Even align_corners is not exposed to upsample.

eellison commented on torch/nn/functional.py:
```diff
@@ -2486,6 +2499,17 @@ def _check_size_scale_factor(dim):
                              'Input is {}D, scale_factor size is {}'.format(dim, len(scale_factor)))
+        if scale_factor is not None:
+            if use_scale_factor is None:
+                is_float_scale_factor = any(not float(scale).is_integer() for scale in _ntuple(dim)(scale_factor))
```
The `float(scale).is_integer()` seems suspicious to me. Why are we not just checking if `scale` is an integer instead of casting it to a float and then checking? I think at least a comment should be added.

lara-hdr replied:
scale.is_integer() would fail if scale is not a float. 
I could have checked 'if scale is int or if scale is float and scale.is_integer()', which seemed to me more complex than a simple call 'float(scale).is_integer()'

eellison replied:
Why are we warning if scale_factor contains `1.0` compared to `1.1`, or if it contains integers? as far as i can tell the function still works regardless:
`[int(math.floor(float(input.size(i + 2)) * scale_factors[i])) for i in range(dim)]`
and
`[int(math.floor(float(input.size(i + 2)) * scale_factors[i])) for i in range(dim)]`

lara-hdr replied:
We are warning if scale_factor contains something like 1.1 but not 1.0.
This was suggested by @fmassa to avoid always printing the warning even when the change will not affect the users.

The output of interpolate will only differ when the scale_factors are floats as explained here https://github.com/pytorch/pytorch/pull/28324#discussion_r355661922.

eellison commented on torch/nn/functional.py:
```diff
@@ -2496,11 +2520,18 @@ def _output_size(dim):
         # make scale_factor a tensor in tracing so constant doesn't get baked in
         if torch._C._get_tracing_state():
-            return [(torch.floor((input.size(i + 2).float() * torch.tensor(scale_factors[i],
+            return [(torch.floor((input.size(i + 2).float() * torch.scalar_tensor(scale_factors[i],
```
Why was this changed to scalar_tensor? Related: I can't find `torch.scalar_tensor` in docs anywhere.

lara-hdr replied:
This is actually unrelated. 
Scalar_tensor will create a tensor as well, which doesn't affect PyTorch, but makes a difference on the ONNX side (tensor is saved a constant in ONNX but not scalar_tensor).
If there is any issues with this line, I could do it in a separate PR.

eellison replied:
yea i guess that would be preferable, mostly just because it's an undocumented function that's not used anywhere else in pytorch library code.

eellison commented on torch/csrc/jit/register_prim_ops.cpp:
```diff
@@ -3355,6 +3355,12 @@ std::vector<int64_t> _output_size(
 
+// return true if v is a real float
+// and false if it is an integer
+bool _is_float_value(float v) {
```
Same comment here: this is suspicious.

eellison replied:
Also TorchScript / Python you a `double` for all floating point values so this should be using `double` as well.

eellison commented on torch/csrc/jit/register_prim_ops.cpp:
```diff
@@ -3382,16 +3389,75 @@ at::Tensor interpolate(
+    bool warn_use_scale_factor = false;
+
+    if (scale_factors.isDouble()) {
+      if(_is_float_value(static_cast<float>(scale_factors.toDouble()))){
```
should be using double here

eellison commented on torch/csrc/jit/register_prim_ops.cpp:
```diff
@@ -3382,16 +3389,75 @@ at::Tensor interpolate(
+    } else if (scale_factors.isDoubleList()) {
+      auto scale_factors_list_ref = scale_factors.toDoubleListRef();
+      std::vector<double> scale_factors_vec(scale_factors_list_ref.begin(), scale_factors_list_ref.end());
+      auto scale_factors_list = c10::impl::toList(scale_factors_vec);
```
`c10::List` api is kind of hard to use, but still we should be just constructing the `c10::List` here instead of constructing both a `std::vector<double>` and a `c10::List`

lara-hdr replied:
I followed the same logic as convert_scale_factor_to_double to construct these lists

eellison commented on torch/csrc/jit/register_prim_ops.cpp:
```diff
@@ -3382,16 +3389,75 @@ at::Tensor interpolate(
+      std::vector<double> scale_factors_vec(scale_factors_list_ref.begin(), scale_factors_list_ref.end());
+      auto scale_factors_list = c10::impl::toList(scale_factors_vec);
+
+      if(_is_float_value(static_cast<float>(scale_factors_list[0]))){
```
These three_if statements could be one for loop

eellison commented on torch/csrc/jit/register_prim_ops.cpp:
```diff
@@ -3382,16 +3389,75 @@ at::Tensor interpolate(
+      scale_factors_2 = static_cast<float>(scale_factors.toDouble());
+      scale_factors_3 = static_cast<float>(scale_factors.toDouble());
+    } else if (scale_factors.isDoubleList()) {
+      auto scale_factors_list_ref = scale_factors.toDoubleListRef();
```
All of the separate handling of the 1, 2, 3 cases is really showing that we should use a `float[]`, the logic would be much simpler.

eellison reviewed (commented):
Commenting mostly on the TorchScript stuff. Generally looks good, but I think we could simplify this PR with a couple changes. 
- I can submit a PR to remove the builtin `interpolate` op and make it so that the normal pytorch code gets compiled. This would remove all the duplicate work you're doing in `register_prim_ops` and make it so that serialized models don't have to get BC'd in upcoming releases.
- As @zou3519 pointed out, it would be nicer if we could use `float[]` here instead. 

I know we said we were targeting `1.4`, but the branch was cut 17 days ago. I don't know how realistic it is to get it in for that release. If we're not, we would have a little more time to do the things I commented on above.

eellison commented on torch/nn/functional.py:
```diff
@@ -2486,6 +2499,17 @@ def _check_size_scale_factor(dim):
                 and len(scale_factor) != dim:
             raise ValueError('scale_factor shape must match input shape. '
                              'Input is {}D, scale_factor size is {}'.format(dim, len(scale_factor)))
+        if scale_factor is not None:
```
nit: `if scale_factor is not None and use_scale_factor is None:`

eellison commented on torch/csrc/jit/register_prim_ops.cpp:
```diff
@@ -3382,16 +3389,75 @@ at::Tensor interpolate(
     }
   }
 
+  float scale_factors_1 = -1.0f;
```
All of `scale_factors_*` should be doubles (or one vector of doubles)

lara-hdr commented:
@eellison , thanks for the review.

> I can submit a PR to remove the builtin interpolate op and make it so that the normal pytorch code gets compiled.

Could we make this change in a following PR?
If I understand correctly this will improve the code, and could be done once we merge this PR.
We are trying to merge this change in 1.4.0, so we can set the desired behavior as default in 1.5.0. 
If we don't merge this in 1.4.0, we will have to wait until 1.6.0 to get the desired default behavior. 

> As @zou3519 pointed out, it would be nicer if we could use float[] here instead.


I agree that this would be better; this was actually the first thing I tried, but using a float[] in the function signatures in the yaml file results in errors and does not compile.

eellison commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -6117,171 +6117,171 @@
     CUDA: upsample_nearest3d_cuda
 
-- func: upsample_nearest3d_backward.grad_input(Tensor grad_output, int[3] output_size, int[5] input_size, *, Tensor(a!) grad_input) -> Tensor(a!)
+- func: upsample_nearest3d_backward.grad_input(Tensor grad_output, int[3] output_size, int[5] input_size, float scales_1=-1.0, float scales_2=-1.0, float scales_3=-1.0, *, Tensor(a!) grad_input) -> Tensor(a!)
```
Is there any situation in which the default values here will be used ? If not, could you remove them ?

lara-hdr replied:
wouldn't it be BC breaking if we add new params without default values?

eellison commented on torch/csrc/jit/symbolic_script.cpp:
```diff
@@ -1261,51 +1270,83 @@ const std::vector<std::string> functions = {
                     align_corners = False
-                grad_self = AD_interpolate_backward(grad_output, input, mode, align_corners)
-                return grad_self, None, None, None, None
+                if use_scale_factor is None:
```
Copy-paste across the four different `interpolate`  would be better factored into a helper function. Helper functions are at the top of this file.

eellison reviewed (commented):
@lara-hdr yea i think it's fine for the fixes to happen in a follow up. 
- When we change a schema from `(*args, float a, float b, float c)` to `(*args, float[])` I don't think it actually breaks backwards compat because pytorch has the variadic list argument matching. Also those functions aren't user-facing. 

- Removing the duplicate bulitin op would be nice but it's not strictly necessary for this pr.
  
However could you try to address / comment on my other review comments ? I was told that PRs against this branch have until the end of the week, so there should be time.

Commit:
address torchscript comments

Commit:
resolve conflict

Commit:
lint

Commit:
Merge branch 'master' of https://github.com/lara-hdr/pytorch into lahaidar/improve_interpolate_onnx

Commit:
nit

Commit:
Merge branch 'master' of https://github.com/lara-hdr/pytorch into lahaidar/improve_interpolate_onnx

eellison commented on torch/csrc/jit/register_prim_ops.cpp:
```diff
@@ -3384,16 +3391,70 @@ at::Tensor interpolate(
     }
   }
 
+  double scale_factors_1 = -1.0f;
```
nit: `-1.0f` -> `1.0`

lara-hdr replied:
good catch. I will wait for the CI to finish running to confirm that there are no new failures, and will then make this fix.

eellison reviewed (approved):
Looks great!!

lara-hdr committed:
remove float annotation from double

lara-hdr commented:
@houseroad is the internal failure related to this PR?

lara-hdr commented:
@houseroad is the internal failure related to this PR?

houseroad commented:
Fix lint issue? https://github.com/pytorch/pytorch/pull/28324/checks?check_run_id=343217323

Commit:
fix lint

eellison commented:
I believe you will also need to submit this PR against v1.4.0 if you want it into 1.4 https://github.com/pytorch/pytorch/issues/30331

lara-hdr commented:
@pytorchbot retest this please

zhangguanheng66 commented:
Have to revert this PR because it breaks the master branch.

houseroad commented:
If it's XLA, we break it on purpose. If it's other branch, please let me know.

zhangguanheng66 commented:
> If it's XLA, we break it on purpose. If it's other branch, please let me know.

Yeap. This is on purpose. Will keep this PR.

lara-hdr commented:
Thank you @houseroad, @ailzhang, @eellison, @fmassa, @zou3519 for helping getting this PR in.
Thanks @houseroad for merging it into 1.4.0.

spandantiwari commented:
Thanks @houseroad @ailzhang @eellison for review and help with merging this.

gchanan commented:
@lara-hdr: can you write a plan english description of what the BC-breaking issue is here?  Something that could go in the release notes, see "Breaking Changes" in one of the releases here: https://github.com/pytorch/pytorch/releases

masahi commented on torch/onnx/symbolic_opset11.py:
```diff
@@ -93,25 +93,37 @@ def pixel_shuffle(g, self, upscale_factor):
+            input_size = g.op("Shape", input)
+            input_size_beg = sym_help._slice_helper(g, input_size, axes=[0], ends=[2], starts=[0])
+            output_size = g.op("Cast", output_size, to_i=sym_help.cast_pytorch_to_onnx["Long"])
+            output_size = g.op("Concat", input_size_beg, output_size, axis_i=0)
```
It seems input_size_beg here is int32, while output_size is int64, can you confirm? @lara-hdr . I'm having a problem converting ONNX Resize op to TVM, because TVM's concat op assumes input dtypes are all same.

lara-hdr commented:
@gchanan, this would be the description:

In Pytorch 1.3, interpolation with scales did not use the scales provided by the user in the calculation of the output; instead it computed the output size that was then used to infer new values for the scales. This processing resulted in precision losses for floating values of scales.
In Pytorch 1.4, we introduced a new attribute use_scale_factor, which allows to directly use the scales provided by the user when enabled.
'use_scale_factor' defaults to False in this release, and the default behavior remains unchanged. Note that the default value will be changed to True in PyTorch 1.5.

Version 1.3
>>> input = torch.arange(4, dtype=torch.float32).reshape(1, 1, 2, 2)
>>> torch.nn.functional.interpolate(input, scale_factor=2.3, mode='bilinear')
tensor([[[[0.0000, 0.2500, 0.7500, 1.0000],
          [0.5000, 0.7500, 1.2500, 1.5000],
          [1.5000, 1.7500, 2.2500, 2.5000],
          [2.0000, 2.2500, 2.7500, 3.0000]]]])

Version 1.4
>>> input = torch.arange(4, dtype=torch.float32).reshape(1, 1, 2 ,2)
>>> torch.nn.functional.interpolate(input, scale_factor=2.3, mode='bilinear', use_scale_factor=True)
tensor([[[[0.0000, 0.1522, 0.5870, 1.0000],
          [0.3043, 0.4565, 0.8913, 1.3043],
          [1.1739, 1.3261, 1.7609, 2.1739],
          [2.0000, 2.1522, 2.5870, 3.0000]]]])

>>> torch.nn.functional.interpolate(input, scale_factor=2.3, mode='bilinear', use_scale_factor=False)		  
tensor([[[[0.0000, 0.2500, 0.7500, 1.0000],
          [0.5000, 0.7500, 1.2500, 1.5000],
          [1.5000, 1.7500, 2.2500, 2.5000],
          [2.0000, 2.2500, 2.7500, 3.0000]]]])

eellison commented on torch/nn/functional.py:
```diff
@@ -2486,6 +2502,16 @@ def _check_size_scale_factor(dim):
                 and len(scale_factor) != dim:
             raise ValueError('scale_factor shape must match input shape. '
                              'Input is {}D, scale_factor size is {}'.format(dim, len(scale_factor)))
+        if scale_factor is not None and use_scale_factor is None:
```
So now that this has been out for a release cycle, what are changes? `use_scale_factor` is set to true ? warnings get removed ? anything else/
